src/Tmds.Ssh/UserAuthentication.PublicKeyAuth.cs

```diff
@@ -224,45 +224,127 @@ byte	3
+                InverseQ = inverseQ.ToByteArray(isUnsigned: true, isBigEndian: true),
+                P = p.ToByteArray(isUnsigned: true, isBigEndian: true),
+                Q = q.ToByteArray(isUnsigned: true, isBigEndian: true),
+                DP = dp.ToByteArray(isUnsigned: true, isBigEndian: true),
+                DQ = dq.ToByteArray(isUnsigned: true, isBigEndian: true)
+            };
+            RSA rsa = RSA.Create();
+            try
+            {
+                rsa.ImportParameters(parameters);
+                privateKey = new RsaPrivateKey(rsa);
+                error = null;
+                return true;
+            }
+            catch (Exception ex)
+            {
+                error = new FormatException($"The data can not be parsed into an RSA key.", ex);
+                rsa.Dispose();
+                return false;
+            }
+        }
+
+        private static bool TryParseOpenSshEcdsaKey(Name keyIdentifier, SequenceReader reader, [NotNullWhen(true)] out PrivateKey? privateKey, [NotNullWhen(false)] out Exception? error)
+        {
+            privateKey = null;
+
+            Name curveName = reader.ReadName();
+
+            HashAlgorithmName allowedHashAlgo;
+            ECCurve curve;
+            if (curveName == AlgorithmNames.Nistp256)
+            {
+                allowedHashAlgo = HashAlgorithmName.SHA256;
+                curve = ECCurve.NamedCurves.nistP256;
+            }
+            else if (curveName == AlgorithmNames.Nistp384)
+            {
+                allowedHashAlgo = HashAlgorithmName.SHA384;
+                curve = ECCurve.NamedCurves.nistP384;
+            }
+            else if (curveName == AlgorithmNames.Nistp521)
+            {
+                allowedHashAlgo = HashAlgorithmName.SHA512;
+                curve = ECCurve.NamedCurves.nistP521;
+            }
+            else
+            {
+                error = new FormatException($"ECDSA curve '{curveName}' is unsupported.");
+                return false;
+            }
+
+            ECPoint q = reader.ReadStringAsECPoint();
+            BigInteger d = reader.ReadMPInt();
+
+            ECDsa ecdsa = ECDsa.Create();
+            try
+            {
+                byte[] dBytes = d.ToByteArray(true, true);
+                if (dBytes.Length < q.X!.Length)
                 {
-                    rsa.ImportParameters(parameters);
-                    privateKey = new RsaPrivateKey(rsa);
-                    error = null;
-                    return true;
+                    Array.Resize(ref dBytes, q.X.Length);
                 }
-                catch (Exception ex)
+
+                ECParameters parameters = new()
                 {
-                    error = new FormatException($"The data can not be parsed into an RSA key.", ex);
-                    rsa?.Dispose();
-                    return false;
+                    Curve = curve,
+                    D = d.ToByteArray(true, true),
```

Comment:
While `ReadMPInt()` reads the raw bytes as a signed value, I found I had to specify the `isUnsigned: true` here or else it would include the signed bit and be too large for `ECParameters`.

Reply:
mpint is defined in https://datatracker.ietf.org/doc/html/rfc4251#section-5. We need to read it as a signed (that is: 2-complement) value.

Reply:
I've added a new overload `ReadMPIntAsBytes()` that returns the `ReadOnlySequence<byte>` which avoiding having to build and then get the bytes of a `BigInteger` value altogether.